Modules/_testclinic.c

```diff
@@ -0,0 +1,960 @@
+/* Always enable assertions */
+#undef NDEBUG
+
+#define PY_SSIZE_T_CLEAN
+
+#include "Python.h"
+
+#include "clinic/_testclinic.c.h"
+
+
+/* Pack arguments to a tuple, implicitly increase all the arguments' refcount.
+ * NULL arguments will be replaced to Py_None. */
+static PyObject *
+pack_arguments_newref(int argc, ...) {
+    assert(!PyErr_Occurred());
+    PyObject *tuple = PyTuple_New(argc);
+    if (!tuple) {
+        return NULL;
+    }
+
+    va_list vargs;
+    va_start(vargs, argc);
+    for (int i = 0; i < argc; i++) {
+        PyObject *arg = va_arg(vargs, PyObject *);
+        if (arg) {
+            if (_PyObject_IsFreed(arg)) {
+                PyErr_Format(PyExc_AssertionError,
+                             "argument %d at %p is freed or corrupted!",
+                             i, arg);
+                va_end(vargs);
+                Py_DECREF(tuple);
+                return NULL;
+            }
+        }
+        else {
+            arg = Py_None;
+        }
+        PyTuple_SET_ITEM(tuple, i, Py_NewRef(arg));
+    }
+    va_end(vargs);
+    return tuple;
+}
+
+/* Pack arguments to a tuple, all the arguments' refcounts should be increased before passing in.
+ * Do not accept NULL arguments unless error occurs. */
+static PyObject *
+pack_arguments(int argc, ...) {
+    assert(!PyErr_Occurred());
+    PyObject *tuple = PyTuple_New(argc);
+    if (!tuple) {
+        return NULL;
+    }
+
+    va_list vargs;
+    va_start(vargs, argc);
+    for (int i = 0; i < argc; i++) {
+        PyObject *arg = va_arg(vargs, PyObject *);
+        assert(arg || PyErr_Occurred());
+        if (!arg) {
+            va_end(vargs);
+            Py_DECREF(tuple);
+            return NULL;
+        }
+        PyTuple_SET_ITEM(tuple, i, arg);
+    }
+    va_end(vargs);
+    return tuple;
+}
+
+
+/*[clinic input]
+module  _testclinic
+[clinic start generated code]*/
+/*[clinic end generated code: output=da39a3ee5e6b4b0d input=d4981b80d6efdb12]*/
+
+
+/*[clinic input]
+test_empty_function
+
+[clinic start generated code]*/
+
+static PyObject *
+test_empty_function_impl(PyObject *module)
+/*[clinic end generated code: output=0f8aeb3ddced55cb input=0dd7048651ad4ae4]*/
+{
+    Py_RETURN_NONE;
+}
+
+
+/*[clinic input]
+objects_converter
+
+    a: object
+    b: object = NULL
+    /
+
+[clinic start generated code]*/
+
+static PyObject *
+objects_converter_impl(PyObject *module, PyObject *a, PyObject *b)
+/*[clinic end generated code: output=3f9c9415ec86c695 input=1533b1bd94187de4]*/
+{
+    return pack_arguments_newref(2, a, b);
+}
+
+
+/*[clinic input]
+bytes_object_converter
+
+    a: PyBytesObject
+    /
+
+[clinic start generated code]*/
+
+static PyObject *
+bytes_object_converter_impl(PyObject *module, PyBytesObject *a)
+/*[clinic end generated code: output=7732da869d74b784 input=94211751e7996236]*/
+{
+    if (!PyBytes_Check(a)) {
+        PyErr_SetString(PyExc_AssertionError,
+                        "argument a is not a PyBytesObject");
+    }
```

Comment:
All functions that raise must return `NULL`. Ditto for all below.

```suggestion
        return NULL;
    }
```